library/bigip_ssl_certificate.py

```diff
@@ -0,0 +1,322 @@
+        return self.api.tm.sys.crypto.certs.cert.exists(name=name)
+
+    def exists_key(self, name):
+        return self.api.tm.sys.crypto.keys.key.exists(name=name)
+
+    def import_cert(self, **kwargs):
+        name = kwargs['name']
+        cert_format = kwargs['cert_format']
+        if cert_format == 'pem':
+            try:
+                certfilename = kwargs['certfilename']
+            except KeyError:
+                certfilename = kwargs['name'] + '.crt'
+            else:
+                self.upload(certfilename)
+                self.install_cert(name, os.path.basename(certfilename))
+            try:
+                keyfilename = kwargs['keyfilename']
+            except KeyError:
+                pass
+            else:
+                self.upload(keyfilename)
+                self.install_key(name, os.path.basename(keyfilename))
+            return True
+        elif cert_format == 'pkcs12':
+            raise F5ModuleError("pkcs12 not supported by f5-sdk")
+
+    def install_cert(self, name, certfilename):
+        filename = os.path.join('/var/config/rest/downloads', certfilename)
+        params = {'from-local-file': filename, 'name': name}
+        self.api.tm.sys.crypto.certs.exec_cmd('install', **params)
+
+    def install_key(self, name, keyfilename):
+        filename = os.path.join('/var/config/rest/downloads', keyfilename)
+        params = {'from-local-file': filename, 'name': name}
+        self.api.tm.sys.crypto.keys.exec_cmd('install', **params)
+
+    def upload(self, filename):
+        self.api.shared.file_transfer.uploads.upload_file(filename)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(type='str', default='present',
+                       choices=['absent', 'present']),
+            user=dict(type='str', required=True, aliases=['username']),
+            password=dict(type='str', required=True, no_log=True),
+            server=dict(type='str', required=True),
+            name=dict(type='str', required=True),
+            partition=dict(type='str', default='Common'),
+            connection=dict(type='str', default='rest',
+                            choices=['rest'],
+                            aliases=['transport']),
+            cert_format=dict(type='str', default='pem',
+                             choices=['pem']),
+            cert_pem_file=dict(type='str'),
+            key_pem_file=dict(type='str'),
+            pkcs12_file=dict(type='str'),
+            pkcs12_password=dict(type='str', no_log=True),
+            validate_cert=dict(type='bool', default=True),
+            # overwrite=dict(type='bool', default=False)
+        )
+    )
+    result = {'changed': False}
+    if module.params['connection'] == 'rest':
+        if not HAS_F5SDK:
+            module.fail_json(msg='connection type rest requires '
+                             'f5-common-python')
+        api = BigIPRest(module.params['server'],
```

Comment:
A question on the flush method approach.  Do we want as much logic as
possible there, or just logic specific to the tasks of interacting with the
F5?

Example for this module, the cert/key filenames are only required if state
is present.  If state is absent, they are not needed, so I did not make
them required in the argument_spec.  Should this logic still be inside of
the class, or should this be handled in main and exit when something is
missing, before we even call the class?

Kevin

On Tue, Jun 14, 2016 at 12:20 PM, Tim Rupp <notifications@github.com> wrote:

> In library/bigip_ssl_certificate.py
> <https://github.com/F5Networks/f5-ansible/pull/18#discussion_r67004277>:
>
> > +            cert_format=dict(type='str', default='pem',
> > +                             choices=['pem']),
> > +            cert_pem_file=dict(type='str'),
> > +            key_pem_file=dict(type='str'),
> > +            pkcs12_file=dict(type='str'),
> > +            pkcs12_password=dict(type='str', no_log=True),
> > +            validate_cert=dict(type='bool', default=True),
> > +            # overwrite=dict(type='bool', default=False)
> > +        )
> > +    )
> > +    result = {'changed': False}
> > +    if module.params['connection'] == 'rest':
> > +        if not HAS_F5SDK:
> > +            module.fail_json(msg='connection type rest requires '
> > +                             'f5-common-python')
> > +        api = BigIPRest(module.params['server'],
>
> if you take a look at line 355 here
>
>
> https://github.com/F5Networks/f5-ansible/blob/master/library/bigip_device_sshd.py#L355
>
> it'd be fantastic if you would follow a similar approach. That approach
> being to keep the execution bits to a minimum and to wrap the logic and
> processing inside of a flush method of an object named after the module.
> In this case, the class name would be BigIpSslCertificate.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/F5Networks/f5-ansible/pull/18/files/c895dea9f2dace29d8951f8b22d541a3109fbe0b#r67004277>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALqwBwAsZicngxyge3tC1hH15Grt69zxks5qLtTEgaJpZM4I0cLm>
> .
>